File: Go/XEnv/variable.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XEnv

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync/atomic"

	"github.com/frameworkease/Universal.X.Utility/Go/XFile"
	"github.com/frameworkease/Universal.X.Utility/Go/XString"
)

type variableScope struct {
	setup     atomic.Bool
	Evaluator XString.IEvaluator
}

type variableEvaluator struct {
	pattern *regexp.Regexp
}

var Variable = &variableScope{
	Evaluator: &variableEvaluator{
		pattern: regexp.MustCompile(`\$\{(?:XEnv\.Variable\.([^}]+?)|([A-Za-z_][A-Za-z0-9_]*))\}`),
	},
}

func (vs *variableScope) ensureSetup() {
	if !vs.setup.Load() {
		vs.Setup()
	}
}

func (vs *variableScope) Setup(extras ...string) {
	// 解析文件
	{
		envFile := filepath.Join(XFile.Directory.Project(), ".env")
		if _, err := os.Stat(envFile); err == nil {
			if data, err := os.ReadFile(envFile); err == nil {
				content := string(data)
				if content != "" {
					lines := strings.SplitSeq(content, "\n")
					for rawLine := range lines {
						line := strings.TrimSpace(rawLine)

						// 跳过空行和注释行
						if line == "" || strings.HasPrefix(line, "#") {
							continue
						}

						// 查找等号分隔符
						before, after, ok := strings.Cut(line, "=")
						if !ok {
							continue
						}

						// 提取键和值
						key := strings.TrimSpace(before)
						value := strings.TrimSpace(after)

						// 跳过无效的键
						if key == "" || value == "" {
							continue
						}

						// 设置环境变量
						os.Setenv(key, value)
					}
				}
			}
		}
	}

	// 解析参数
	{
		for _, extra := range extras {
			if extra == "" {
				continue
			}

			// 解析 "KEY=VALUE" 格式
			before, after, ok := strings.Cut(extra, "=")
			if !ok {
				continue
			}

			key := strings.TrimSpace(before)
			value := strings.TrimSpace(after)

			// 跳过无效的键或值
			if key == "" || value == "" {
				continue
			}

			// 设置环境变量
			os.Setenv(key, value)
		}
	}
	vs.setup.Store(true)
}

func (vs *variableScope) Get(key string) string {
	vs.ensureSetup()
	return os.Getenv(key)
}

func (vs *variableScope) Exists(key string) bool {
	vs.ensureSetup()
	val, ok := os.LookupEnv(key)
	return val != "" && ok
}

func (vs *variableScope) Range(callback func(key, value string) bool) {
	vs.ensureSetup()

	if callback == nil {
		panic("XEnv.Variable.Range: callback is nil.")
	}
	for _, env := range os.Environ() {
		before, after, ok := strings.Cut(env, "=")
		if ok {
			if !callback(before, after) {
				break
			}
		}
	}
}

func (ve *variableEvaluator) Evaluate(expr string) string {
	if expr == "" {
		return expr
	}

	visited := make(map[string]bool)

	var repl func(string) string
	repl = func(m string) string {
		sm := ve.pattern.FindStringSubmatch(m)
		g1, g2 := sm[1], sm[2]

		var key string
		var value string
		if g2 != "" {
			key = "Variable." + g2
			value = Variable.Get(g2)
		} else {
			key = "Variable." + g1
			if strings.Contains(g1, "${") {
				return fmt.Sprintf("${Nested.%v}", m)
			}
			value = Variable.Get(g1)
		}

		if visited[key] {
			return fmt.Sprintf("${Recursive.XEnv.%v}", key)
		}
		visited[key] = true
		defer delete(visited, key)

		if value != "" {
			return ve.pattern.ReplaceAllStringFunc(value, repl)
		}
		if g2 != "" {
			return fmt.Sprintf("${Unknown.XEnv.Variable.%v}", g2)
		}
		return fmt.Sprintf("${Unknown.XEnv.Variable.%v}", g1)
	}

	return ve.pattern.ReplaceAllStringFunc(expr, repl)
}

File: CSharp/src/Runtime/Scripts/XEnv/variable.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XEnv
    {
        public static class Variable
        {
            internal static volatile bool setup;

            internal static readonly VariableEvaluator evaluator = new();

            public static XString.IEvaluator Evaluator => evaluator;

            internal static void EnsureSetup() { if (!setup) Setup(); }

            public static void Setup(params string[] extras)
            {
                #region 解析文件
                try
                {
                    var envFile = Path.Join(XFile.Directory.Project, ".env");
                    if (XFile.Exists(envFile))
                    {
                        var lines = File.ReadAllLines(envFile);
                        foreach (var rawLine in lines)
                        {
                            var line = rawLine.Trim();
                            if (string.IsNullOrWhiteSpace(line) || line.StartsWith('#')) continue;

                            var idx = line.IndexOf('=');
                            if (idx < 0) continue;

                            var key = line[..idx].Trim();
                            var value = line[(idx + 1)..].Trim();
                            if (string.IsNullOrWhiteSpace(key) || string.IsNullOrWhiteSpace(value)) continue;

                            Environment.SetEnvironmentVariable(key, value);
                        }
                    }
                }
                catch (Exception e) { XLog.Error(e); }
                #endregion

                #region 解析参数
                if (extras is { Length: > 0 })
                {
                    foreach (var extra in extras)
                    {
                        if (string.IsNullOrEmpty(extra)) continue;

                        var idx = extra.IndexOf('=');
                        if (idx < 0) continue;

                        var key = extra[..idx].Trim();
                        var value = extra[(idx + 1)..].Trim();
                        if (string.IsNullOrWhiteSpace(key) || string.IsNullOrWhiteSpace(value)) continue;

                        Environment.SetEnvironmentVariable(key, value);
                    }
                }
                #endregion

                setup = true;
            }

            public static string Get(string key)
            {
                EnsureSetup();
                return Environment.GetEnvironmentVariable(key);
            }

            public static bool Exists(string key)
            {
                EnsureSetup();
                return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key));
            }

            public static void Range(Func<string, string, bool> callback)
            {
                EnsureSetup();
                if (callback == null) throw new Exception("XEnv.Variable.Range: callback is null.");
                foreach (System.Collections.DictionaryEntry entry in Environment.GetEnvironmentVariables())
                {
                    var key = entry.Key.ToString();
                    var value = entry.Value?.ToString() ?? string.Empty;
                    if (!callback(key, value)) break;
                }
            }
        }

        internal sealed class VariableEvaluator : XString.IEvaluator
        {
            internal static readonly Regex Pattern = new(@"\$\{(?:XEnv\.Variable\.([^}]+?)|([A-Za-z_][A-Za-z0-9_]*))\}", RegexOptions.Compiled);

            public string Evaluate(string expr)
            {
                if (string.IsNullOrEmpty(expr)) return expr;
                var visited = new HashSet<string>();

                string repl(Match m)
                {
                    var g1 = m.Groups[1].Value;
                    var g2 = m.Groups[2].Value;

                    string key;
                    string value = null;

                    if (!string.IsNullOrEmpty(g2))
                    {
                        key = "Variable." + g2;
                        value = Variable.Get(g2);
                    }
                    else
                    {
                        key = "Variable." + g1;
                        if (g1.Contains("${")) return $"${{Nested.{m.Value}}}";
                        value = Variable.Get(g1);
                    }

                    if (!visited.Add(key)) return $"${{Recursive.XEnv.{key}}}";
                    try
                    {
                        if (!string.IsNullOrEmpty(value)) return Pattern.Replace(value, repl);
                        if (!string.IsNullOrEmpty(g2)) return $"${{Unknown.XEnv.Variable.{g2}}}";
                        return $"${{Unknown.XEnv.Variable.{g1}}}";
                    }
                    finally { visited.Remove(key); }
                }

                return Pattern.Replace(expr, repl);
            }
        }
    }
}
